File: packages/rank/ios/Sources/RankPlugin/RankPlugin.swift
import Foundation
import Capacitor

/**
 * Capacitor bridge for the Rank plugin.
 *
 * This class handles the communication between the JavaScript layer and the native iOS implementation.
 * It is responsible for input validation, configuration merging, and thread safety.
 */
@objc(RankPlugin)
public final class RankPlugin: CAPPlugin, CAPBridgedPlugin {

    // MARK: - Properties

    /// An instance of the implementation class that contains the plugin's core functionality.
    private let implementation = RankImpl()

    /// Internal storage for the plugin configuration read from capacitor.config.ts.
    private var config: RankConfig?

    /// The unique identifier for the plugin used by the Capacitor bridge.
    public let identifier = "RankPlugin"

    /// The name used to reference this plugin in JavaScript (e.g., Rank.requestReview()).
    public let jsName = "Rank"

    /**
     * A list of methods exposed by this plugin to the JavaScript layer.
     * All methods defined here must be implemented with the @objc attribute.
     - `isAvailable`: Checks if the native review prompt can be shown on the current device.
     - `requestReview`: Triggers the native iOS In-App Review prompt.
     - `presentProductPage`: Navigates the user to the App Store product page within the app.
     - `openStore`: Opens the app's page on the App Store for leaving a review or viewing details.
     - `getPluginVersion`: A method that returns the version of the plugin.
     */
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkReviewEnvironment", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestReview", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "presentProductPage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openStore", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openStoreListing", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "search", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openDevPage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openCollection", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    // MARK: - Lifecycle

    /**
     * Plugin lifecycle entry point.
     *
     * Called once when the plugin is loaded. This method initializes the configuration
     * and prepares the native implementation.
     */
    override public func load() {
        // Initialize RankConfig with the correct type
        let cfg = RankConfig(plugin: self)
        self.config = cfg
        implementation.applyConfig(cfg)

        // Log if verbose logging is enabled
        RankLogger.debug("Plugin loaded.")
    }

    // MARK: - Error Mapping

    /**
     * Rejects the call using standardized error codes from the native RankError enum.
     */
    private func reject(
        _ call: CAPPluginCall,
        error: RankError
    ) {
        // Use the centralized errorCode and message defined in RankError.swift
        call.reject(error.message, error.errorCode)
    }

    private func handleError(_ call: CAPPluginCall, _ error: Error) {
        if let rankError = error as? RankError {
            call.reject(rankError.message, rankError.errorCode)
        } else {
            reject(call, error: .initFailed(error.localizedDescription))
        }
    }

    // MARK: - Availability

    /**
     * Checks if the native In-App Review prompt can be displayed on the current device.
     *
     * This method verifies if the necessary APIs are available and if the OS version is compatible.
     * It returns a boolean indicating availability, which can be used by the JavaScript layer to
     * conditionally show review prompts or fallback UI.
     *
     * - Parameter call: CAPPluginCall used to return the availability result.
     */
    @objc func isAvailable(_ call: CAPPluginCall) {
        let available = implementation.isAvailable()
        call.resolve([
            "value": available
        ])
    }

    @objc func checkReviewEnvironment(_ call: CAPPluginCall) {
        call.resolve([
            "canRequestReview": false,
            "reason": "PLAY_STORE_NOT_AVAILABLE"
        ])
    }

    // MARK: - Product Page

    /**
     * Presents the App Store product page within the application.
     *
     * This method uses SKStoreProductViewController to display the product page without leaving the app.
     * It accepts an optional appId parameter, which can override the global configuration.
     *
     * - Parameter call: CAPPluginCall containing:
     * - appId (String, optional): The Apple App ID. Fallbacks to global config.
     */
    @objc func presentProductPage(_ call: CAPPluginCall) {
        let rawAppId = call.getString("appId") ?? config?.appleAppId
        guard let appId = RankValidators.validateAppId(rawAppId) else {
            reject(call, error: .invalidInput(RankErrorMessages.invalidAppleAppId))
            return
        }

        implementation.presentProductPage(appId: appId) { success, error in
            // Ensure bridge resolution occurs on the main thread.
            DispatchQueue.main.async {
                if success {
                    call.resolve()
                } else {
                    if let rankError = error as? RankError {
                        self.reject(call, error: rankError)
                    } else {
                        self.reject(call, error: .initFailed(error?.localizedDescription ?? RankErrorMessages.productPageLoadFailed))
                    }
                }
            }
        }
    }

    // MARK: - Public Plugin Methods

    /**
     * Triggers the native iOS In-App Review prompt using SKStoreReviewController.
     *
     * This method resolves immediately as iOS does not provide a callback to determine
     * if the prompt was shown or if the user interacted with it.
     *
     * - Parameter call: CAPPluginCall provided by the bridge.
     */
    @objc func requestReview(_ call: CAPPluginCall) {
        // Log the request if verbose logging is enabled
        RankLogger.debug("Requesting review prompt.")

        // Native iOS SKStoreReviewController logic
        implementation.requestReview()

        // Always resolve immediately for iOS consistency
        call.resolve()
    }

    // MARK: - Store Navigation

    /**
     * Navigates the user to the App Store page for a specific application.
     *
     * - Parameter call: CAPPluginCall containing:
     * - appId (String, optional): The Apple App ID. Fallbacks to global config.
     */
    @objc func openStore(_ call: CAPPluginCall) {
        let rawAppId = call.getString("appId") ?? config?.appleAppId
        // Attempt to retrieve the App ID from call parameters or static configuration
        guard let appId = RankValidators.validateAppId(rawAppId) else {
            reject(call, error: .invalidInput(RankErrorMessages.invalidAppleAppId))
            return
        }

        RankLogger.debug("Opening App Store for ID: \(appId)")
        implementation.openStore(appId: appId)
        call.resolve()
    }

    /**
     * Opens a collection page on the App Store.
     *
     * This method is not supported on iOS as there is no direct way to link to collections. It will reject with an appropriate message.
     *
     * - Parameter call: CAPPluginCall provided by the bridge.
     */
    @objc func openCollection(_ call: CAPPluginCall) {
        reject(call, error: .unavailable(RankErrorMessages.collectionsNotSupportedIos))
    }

    /**
     * Opens the App Store page for the specified application ID.
     *
     * This method constructs a deep link using the `itms-apps` scheme to direct the user
     * to the review section of the application's store page.
     *
     * - Parameter appId: The numeric Apple App ID (e.g., "123456789").
     */
    @objc func openStoreListing(_ call: CAPPluginCall) {
        let rawAppId = call.getString("appId") ?? config?.appleAppId
        guard let appId = RankValidators.validateAppId(rawAppId) else {
            reject(call, error: .invalidInput(RankErrorMessages.invalidAppleAppId))
            return
        }
        implementation.openStoreListing(appId: appId)
        call.resolve()
    }

    // MARK: - Search

    /**
     * Searches the App Store for the given terms.
     *
     * - Parameter call: CAPPluginCall containing:
     * - terms (String): The search query to find apps or developers.
     */
    @objc func search(_ call: CAPPluginCall) {
        guard let terms = call.getString("terms") else {
            reject(call, error: .invalidInput(RankErrorMessages.invalidSearchTerms))
            return
        }
        implementation.search(terms: terms)
        call.resolve()
    }

    // MARK: - Developer Page

    /**
     * Opens the developer's page on the App Store.
     *
     * Since iOS does not have a direct link for developer pages, this method performs a search using the developer's name or ID.
     *
     * - Parameter call: CAPPluginCall containing:
     * - devId (String): The developer identifier or name to search for.
     */
    @objc func openDevPage(_ call: CAPPluginCall) {
        guard let devId = call.getString("devId") else {
            reject(call, error: .invalidInput(RankErrorMessages.invalidDeveloperId))
            return
        }
        implementation.search(terms: devId) // Fallback on iOS
        call.resolve()
    }

    // MARK: - Version

    /**
     * Retrieves the current native plugin version.
     *
     * This version is synchronized from the project's package.json during the build process.
     *
     * - Parameter call: CAPPluginCall used to return the version string.
     */
    @objc func getPluginVersion(_ call: CAPPluginCall) {
        // Standardized enum name across all CapKit plugins
        call.resolve([
            "version": PluginVersion.number
        ])
    }
}

File: packages/rank/android/src/main/java/io/capkit/rank/RankPlugin.kt
package io.capkit.rank

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import io.capkit.rank.error.RankErrorMessages
import io.capkit.rank.utils.RankLogger
import io.capkit.rank.utils.RankValidators

/**
 * Capacitor bridge for the Rank plugin.
 *
 * This class acts as the boundary between JavaScript and native Android code.
 * It handles input parsing, configuration management, and delegates execution
 * to the platform-specific implementation.
 */
@CapacitorPlugin(name = "Rank")
class RankPlugin : Plugin() {
  // ---------------------------------------------------------------------------
  // Properties
  // ---------------------------------------------------------------------------

  /**
   * Immutable plugin configuration read from capacitor.config.ts.
   * * CONTRACT:
   * - Initialized exactly once in `load()`.
   * - Treated as read-only afterwards.
   */
  private lateinit var config: RankConfig

  /**
   * Native implementation layer containing core Android logic.
   *
   * CONTRACT:
   * - Owned by the Plugin layer.
   * - MUST NOT access PluginCall or Capacitor bridge APIs directly.
   */
  private lateinit var implementation: RankImpl

  // ---------------------------------------------------------------------------
  // Lifecycle
  // ---------------------------------------------------------------------------

  /**
   * Called once when the plugin is loaded by the Capacitor bridge.
   *
   * This method initializes the configuration container and the native
   * implementation layer, ensuring all dependencies are injected.
   */
  override fun load() {
    super.load()

    config = RankConfig(this)
    implementation = RankImpl(context)
    implementation.updateConfig(config)

    RankLogger.verbose = config.verboseLogging
    RankLogger.debug("Plugin loaded")

    // RULE: Perform pre-warm of native resources to improve UX
    implementation.preloadReviewInfo()
  }

  // ---------------------------------------------------------------------------
  // Error Mapping
  // ---------------------------------------------------------------------------

  /**
   * Rejects the call with a message and a standardized error code.
   * Ensure consistency with the JS RankErrorCode enum.
   */
  private fun reject(
    call: PluginCall,
    error: RankError,
  ) {
    val code =
      when (error) {
        is RankError.Unavailable -> "UNAVAILABLE"
        is RankError.Cancelled -> "CANCELLED"
        is RankError.PermissionDenied -> "PERMISSION_DENIED"
        is RankError.InitFailed -> "INIT_FAILED"
        is RankError.InvalidInput -> "INVALID_INPUT"
        is RankError.UnknownType -> "UNKNOWN_TYPE"
        is RankError.NotFound -> "NOT_FOUND"
        is RankError.Conflict -> "CONFLICT"
        is RankError.Timeout -> "TIMEOUT"
      }

    // Always use the message from the RankError instance
    val message = error.message ?: "Unknown native error"
    call.reject(message, code)
  }

  // ---------------------------------------------------------------------------
  // Availability
  // ---------------------------------------------------------------------------

  /**
   * Checks if the device supports the Google Play Review API.
   */
  @PluginMethod
  fun isAvailable(call: PluginCall) {
    implementation.isAvailable { available ->
      val ret = JSObject()
      ret.put("value", available)
      call.resolve(ret)
    }
  }

  /**
   * Performs a diagnostic check to determine whether the Google Play
   * In-App Review dialog can be displayed in the current environment_attach.
   *
   * This method does NOT trigger the review flow and does NOT represent
   * an error condition. It is intended purely for runtime diagnostics.
   *
   * Typical cases where the review dialog cannot be shown include:
   * - The Google Play Store app is not installed or is disabled
   * - The app was not installed from the official Play Store
   * - The device or environment is not Play Store–certified (e.g. emulator)
   *
   * The result is returned as a structured object instead of rejecting
   * the call, allowing consumers to make UI decisions (e.g. disabling
   * the review button) without treating this as a failure.
   */
  @PluginMethod
  fun checkReviewEnvironment(call: PluginCall) {
    implementation.checkReviewEnvironment { diagnostic ->
      if (diagnostic.error != null) {
        reject(call, diagnostic.error)
        return@checkReviewEnvironment
      }

      val ret = JSObject()
      ret.put("canRequestReview", diagnostic.canRequestReview)

      if (!diagnostic.canRequestReview && diagnostic.reason != null) {
        ret.put("reason", diagnostic.reason)
      }

      call.resolve(ret)
    }
  }

  // ---------------------------------------------------------------------------
  // Product Page (Fallback for Android)
  // ---------------------------------------------------------------------------

  /**
   * Android fallback for presentProductPage.
   * Redirects to the Play Store app as no native internal overlay exists.
   */
  @PluginMethod
  fun presentProductPage(call: PluginCall) {
    openStore(call)
  }

  // ---------------------------------------------------------------------------
  // Public Plugin Methods
  // ---------------------------------------------------------------------------

  /**
   * Requests the display of the native Google Play In-App Review flow.
   *
   * This method extracts the 'fireAndForget' policy to determine if the
   * promise should resolve immediately or wait for the review flow results.
   *
   * * @param call PluginCall containing:
   * - fireAndForget (Boolean, optional): Override for the global resolution policy.
   */
  @PluginMethod
  fun requestReview(call: PluginCall) {
    val fireAndForget = call.getBoolean("fireAndForget") ?: config.fireAndForget

    // If fireAndForget is enabled, resolve the call immediately
    if (fireAndForget) {
      call.resolve()
    }

    val currentActivity = activity
    if (currentActivity == null) {
      // Edge case: plugin invoked while Activity is not available
      RankLogger.error("Cannot request review: Activity is null")
      if (!fireAndForget) {
        reject(call, RankError.Unavailable(RankErrorMessages.ACTIVITY_NOT_AVAILABLE))
      }
      return
    }

    // RULE: UI operations MUST be executed on the main thread
    currentActivity.runOnUiThread {
      implementation.requestReview(currentActivity) { error: RankError? ->
        // Ensure we only respond if we haven't already resolved via fireAndForget
        if (!fireAndForget) {
          if (error != null) {
            RankLogger.error("Review flow failed", error)
            reject(call, error)
          } else {
            call.resolve()
          }
        }
      }
    }
  }

  /**
   * Navigates the user to the Google Play Store page for the application.
   *
   * * @param call PluginCall containing:
   * - packageName (String, optional): The target app package. Defaults to the host app.
   */
  @PluginMethod
  fun openStore(call: PluginCall) {
    val rawPackageName = call.getString("packageName") ?: config.androidPackageName
    val packageName = RankValidators.validatePackageName(rawPackageName)

    if (packageName == null) {
      reject(call, RankError.InvalidInput(RankErrorMessages.INVALID_ANDROID_PACKAGE_NAME))
      return
    }

    val currentActivity = activity
    if (currentActivity == null) {
      reject(call, RankError.Unavailable(RankErrorMessages.ACTIVITY_NOT_AVAILABLE))
      return
    }

    currentActivity.runOnUiThread {
      try {
        implementation.openStore(packageName)
        call.resolve()
      } catch (e: Exception) {
        reject(call, RankError.InitFailed(e.message ?: RankErrorMessages.NATIVE_OPERATION_FAILED))
      }
    }
  }

  /**
   * Opens the App Store listing page.
   * * This method resolves the target identifier by checking the 'appId' parameter,
   * then 'packageName', and finally falling back to the configured static ID.
   * * @param call PluginCall containing optional 'appId' or 'packageName'.
   */
  @PluginMethod
  fun openStoreListing(call: PluginCall) {
    val rawPackageName =
      call.getString("appId")
        ?: call.getString("packageName")
        ?: config.androidPackageName

    val packageName = RankValidators.validatePackageName(rawPackageName)

    if (packageName == null) {
      reject(call, RankError.InvalidInput(RankErrorMessages.INVALID_ANDROID_PACKAGE_NAME))
      return
    }

    val currentActivity = activity
    if (currentActivity == null) {
      reject(call, RankError.Unavailable(RankErrorMessages.ACTIVITY_NOT_AVAILABLE))
      return
    }

    currentActivity.runOnUiThread {
      try {
        implementation.openStoreListing(packageName)
        call.resolve()
      } catch (e: Exception) {
        reject(call, RankError.InitFailed(e.message ?: RankErrorMessages.NATIVE_OPERATION_FAILED))
      }
    }
  }

  /**
   * Opens a specific app collection on the Play Store.
   * * @param call PluginCall containing the required 'name' string.
   */
  @PluginMethod
  fun openCollection(call: PluginCall) {
    val rawName = call.getString("name")
    val name = RankValidators.validateCollectionName(rawName)

    if (name == null) {
      reject(call, RankError.InvalidInput(RankErrorMessages.INVALID_COLLECTION_NAME))
      return
    }

    val currentActivity = activity
    if (currentActivity == null) {
      reject(call, RankError.Unavailable(RankErrorMessages.ACTIVITY_NOT_AVAILABLE))
      return
    }

    currentActivity.runOnUiThread {
      implementation.openCollection(name)
      call.resolve()
    }
  }

  /**
   * Performs a store search for the given terms.
   * * @param call PluginCall containing the required 'terms' string.
   */
  @PluginMethod
  fun search(call: PluginCall) {
    val rawTerms = call.getString("terms")
    val terms = RankValidators.validateSearchTerms(rawTerms)

    if (terms == null) {
      reject(call, RankError.InvalidInput(RankErrorMessages.INVALID_SEARCH_TERMS))
      return
    }

    val currentActivity = activity
    if (currentActivity == null) {
      reject(call, RankError.Unavailable(RankErrorMessages.ACTIVITY_NOT_AVAILABLE))
      return
    }

    currentActivity.runOnUiThread {
      implementation.search(terms)
      call.resolve()
    }
  }

  /**
   * Opens the developer page on the store.
   * * @param call PluginCall containing the required 'devId' string.
   */
  @PluginMethod
  fun openDevPage(call: PluginCall) {
    val rawDevId = call.getString("devId")
    val devId = RankValidators.validateDevId(rawDevId)

    if (devId == null) {
      reject(call, RankError.InvalidInput(RankErrorMessages.INVALID_DEVELOPER_ID))
      return
    }

    val currentActivity = activity
    if (currentActivity == null) {
      reject(call, RankError.Unavailable(RankErrorMessages.ACTIVITY_NOT_AVAILABLE))
      return
    }

    currentActivity.runOnUiThread {
      implementation.openDevPage(devId)
      call.resolve()
    }
  }

  // ---------------------------------------------------------------------------
  // Version Information
  // ---------------------------------------------------------------------------

  /**
   * Returns the native plugin version synchronized from package.json.
   *
   * This information is used for diagnostics and ensuring parity between
   * the JavaScript and native layers.
   *
   * @param call The bridge call to resolve with version data.
   */
  @PluginMethod
  fun getPluginVersion(call: PluginCall) {
    val ret = JSObject()
    ret.put("version", BuildConfig.PLUGIN_VERSION)
    call.resolve(ret)
  }
}
